runtime/datastructures/src/decls.rs

```diff
@@ -69,6 +101,86 @@ impl Hash for KElem {
   }
 }
 
+fn hash_map_compare<K,V>(a: &HashMap<K, V>, b: &HashMap<K, V>) -> Ordering
+where K: Ord + Hash + Clone, V: Ord + Clone {
+    let mut a_keys : BinaryHeap<_> = a.keys().collect();
+    let mut b_keys : BinaryHeap<_> = b.keys().collect();
+    while !(a_keys.is_empty() || b_keys.is_empty()) {
+        let (a_top, b_top) = (a_keys.pop(), b_keys.pop());
+        if a_top > b_top {
+            return Ordering::Greater;
+        }
+        if b_top > a_top {
+            return Ordering::Less;
+        }
+        let (a_top_elem, b_top_elem) = (a.get(a_top.unwrap()), b.get(b_top.unwrap()));
+        if a_top_elem > b_top_elem {
+            return Ordering::Greater;
+        }
+        if b_top_elem > a_top_elem {
+            return Ordering::Less;
+        }
+    }
+    if a_keys.is_empty() {
+        if b_keys.is_empty() {
+            return Ordering::Equal;
+        }
+        return Ordering::Less;
+    }
+    return Ordering::Greater;
+}
+
+fn hash_set_compare<K>(a: &HashSet<K>, b: &HashSet<K>) -> Ordering
+where K: Ord + Hash + Clone {
+    let mut a_vals : BinaryHeap<_> = a.iter().collect();
+    let mut b_vals : BinaryHeap<_> = b.iter().collect();
+    while !(a_vals.is_empty() || b_vals.is_empty()) {
+        let (a_top, b_top) = (a_vals.pop(), b_vals.pop());
+        if a_top > b_top {
+            return Ordering::Greater;
+        }
+        if b_top > a_top {
+            return Ordering::Less;
+        }
+    }
+    if a_vals.is_empty() {
+        if b_vals.is_empty() {
+            return Ordering::Equal;
+        }
+        return Ordering::Less;
+    }
+    return Ordering::Greater;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_LIST_cmp(a: *const c_void, b: *const c_void) -> i64 {
```

Comment:
Don't use c_void here, it's not actually necessary. Just use *const List. Then you need to transmute and it just becomes part of the abi of the call.

Reply:
Interestingly, I figured this out independently.